### PR TITLE
[Diffusion] Enable request-level batching for Boogu-Image (TI2I)

### PR DESCRIPTION
**Pull request by ShengleiFu**
## Summary

Enable request-level batching for Boogu-Image TI2I (follow-up to #6968; tracks #6703 / #6665 Track 2).

- Allow compatible TI2I requests to batch while keeping T2I and TI2I isolated.
- Forward `guidance_scale_2` in the edit benchmark and retain opt-in failure-detail logging.
- Add partner-isolation tests for prompt, negative prompt, seed, and reference image; injected mixing bugs validate the test controls.

## Validation

- CPU: 76 tests passed in the author's run of `tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py`.
- GPU correctness: fixed-shape B=2 partner-isolation checks were bit-exact for the tested image-only and double-guidance cases. Different partner instruction lengths can still drift (MAD 2.6–6.0); this is outside the bit-exact claim.
- GPU performance: RTX PRO 6000, `Boogu-Image-0.1-Edit`, TI2I, 512×512, 20 steps, guidance 5/2, `max_num_seqs=1` vs 2, zero batch wait.

### Performance

Each server received a 32-request full-workload warmup. Measured arms used the intended guidance values, recorded actual forward batch sizes, and had no new encoded shapes or TorchDynamo graphs. For the five saturation pairs, each arm also started only after three 10-second-spaced samples at or below 55°C; order alternated across pairs.

| Load | Serial imgs/s | B2 imgs/s | Delta | Actual B2 occupancy |
|---|---:|---:|---:|---|
| Saturation, 5 cooled paired runs × 32 requests/arm | 0.15273 mean | 0.16622 mean | **+8.87%** mean paired gain (range +6.52% to +11.05%) | 30/32 requests per B2 arm |
| Offered QPS 0.10, 1 cooled pair × 14 requests/arm | 0.14982 | 0.15122 | +0.93% | 4/14 requests (2 B2 forwards) |
| Offered QPS 0.15, 1 cooled pair × 14 requests/arm | 0.15766 | 0.16670 | +5.74% | 12/14 requests (6 B2 forwards) |

Low-QPS benefit depends on actual co-arrival: at offered QPS 0.10, few requests formed B2, so the throughput difference is near zero. The finite-QPS rows are single-pair screening results; the five-pair saturation result is the variance-controlled estimate.

## AI assistance

AI-assisted implementation (Claude), with review and benchmark/documentation cleanup assisted by Codex.

### CODE REVIEW

**Comment by ShengleiFu** [comment]
Self-review completed.
I reviewed the request-batching path and benchmark changes, with particular attention to:
- TI2I-only compatibility and T2I/TI2I isolation.
- Per-request seed/reference-image isolation in batched execution.
- guidance_scale_2 forwarding in the edit benchmark and opt-in failure logging.
- Unit-test controls that detect intentional seed/reference mixing.
I also re-ran GPU validation for TI2I batching: full-workload warmup showed no measured-arm recompilation, and cooled paired saturation runs showed a consistent B2 throughput gain.

**Comment by yenuo26** [comment]
Boogu Edit's double-guidance uses a three-branch CFG path, but there doesn't seem to be any evidence in the validation results for the combination of cfg_parallel_size=2/3 with TI2I B=2.

File: tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py [hunk 1]
```diff
@@ -1317,25 +1336,20 @@ def test_forward_image_guidance_ignored_without_reference():
 # ---------------------------------------------------------------------------
 
 
-def test_boogu_batch_compatibility_key_t2i_stable_ti2i_unique():
+def test_boogu_batch_compatibility_key_t2i_and_ti2i_stable_and_isolated():
```
Comment (yenuo26) [inline tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py:1340]: This test does not protect the reason TI2I was gated to batch=1. Do we need to assert that two TI2I `OmniDiffusionRequest`s which differ only in `guidance_scale_2_provided` get unequal `build_request_batch_sampling_params_key` results.? Dropping that field from `RequestBatchSamplingParamsKey` would still leave this CPU test green.

**Comment by yenuo26** [comment]
@vllm-omni-review-bot

**Comment by hsliuustc0106** [comment]
Reviewed `acbef19b63f63fb9cd9e238ebd6b1f998e31d451`.

I found no concrete correctness defect in the TI2I batch enablement in this snapshot. The scheduler key includes `guidance_scale_2_provided`, and `tests/diffusion/test_diffusion_scheduler.py` already contains `test_distinguishes_explicit_guidance_scale_2`, which already checks omitted versus explicitly supplied guidance at the scheduler boundary. I would not repeat this as a missing-coverage finding.

The paired throughput runs and partner-isolation controls are useful. Remaining validation is the supported combination of TI2I B=2 with CFG parallelism, plus broader prompt-length/reference-shape cases. Keep the bit-exact claim restricted to the fixed-shape cases actually tested.

Validation: static source/test review and available CI evidence; I did not run fork code or reproduce GPU benchmarks.

**Comment by hsliuustc0106** [comment]
Local performance validation on `acbef19b63f63fb9cd9e238ebd6b1f998e31d451`: enabling request batching gave **+16.25% mean paired saturated throughput** on one NVIDIA L20X, with a **16.16–16.38%** range across three pairs.

Both arms used this PR head, with `--max-num-seqs 1` versus `2` and `--request-batch-max-wait-ms 0`. This measures the batching benefit on the same code revision, rather than an upstream-base/head regression comparison.

Setup: Boogu/Boogu-Image-0.1-Edit at model revision `0cce2581de53bfdb5d67aa438e39c3d6e7556808`, BF16 native pipeline, FlashAttention, default regional `torch.compile`, no profiler. Python 3.12.13, PyTorch 2.13.0+cu130, vLLM 0.28.0, Transformers 5.14.1, Diffusers 0.40.0; driver 570.133.20.

Workload: one fixed synthetic 512×512 house reference image, instruction “Change the style to a colored pencil drawing.”, 20 steps, text/image guidance 5/2, seeds 42–73. Each arm received 32 full-workload warmup requests. Measured bursts contained 32 requests at concurrency 32; pair order was B1/B2, B2/B1, B1/B2. Both servers remained resident on the same GPU, but only one received requests at a time. Before each measured arm, three samples spaced 10,000 ms apart required temperature ≤55°C and GPU utilization ≤2%.

| Pair | B1 images/s | B2 images/s | Gain |
|---|---:|---:|---:|
| 1 | 0.22452 | 0.26129 | +16.38% |
| 2 | 0.22519 | 0.26157 | +16.16% |
| 3 | 0.22486 | 0.26133 | +16.22% |

An identical one-line log at pipeline dispatch confirmed that every B2 burst executed 15 two-request forwards plus two singleton forwards: **30/32 requests batched**. All B1 forwards were singletons.

| Metric | B1 | B2 |
|---|---:|---:|
| Mean 32-request burst duration | 142,313 ms | 122,419 ms |
| Mean queued request latency | 73,472 ms | 64,960 ms |
| Mean per-burst p95 request latency | 137,844 ms | 117,889 ms |
| Isolated-request mean latency, 3 requests | 4,480 ms | 4,562 ms |
| Peak per-process CUDA reserved memory | 36,892 MiB | 37,826 MiB |

All 198 measured requests succeeded. The isolated-request sample shows no single-request speedup; the benefit here is saturated throughput. Burst latencies include queueing.

**Numerical parity caveat:** comparing 32 matching seeds across B1/B2, only 2/32 outputs were pixel-exact. Mean pixel MAE was **1.70/255**, maximum per-image MAE **2.68/255**, and minimum PSNR **34.35 dB**. Comparing the first two measured runs within each configuration, B1 repeated exactly for 32/32 images and B2 repeated exactly for 32/32 images. These are serial-versus-batched output checks, distinct from the PR's fixed-batch partner-isolation claim.

This supports the throughput gain for this fixed-shape workload, but does not establish exact serial/batch parity or quality across varying reference images, prompt lengths, or low arrival rates.

File: tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py (hunk 1 above)
Comment (ShengleiFu) [reply tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py:1340]: Good point, added the test, PTAL, thanks!

**Comment by ShengleiFu** [comment]
Validated the requested full-model TI2I request-batching interaction with CFG parallelism on PR head `acbef19b`.

Configuration: Boogu-Image Edit, request batch B=2, 512×512, 20 steps, `guidance_scale=5`, `guidance_scale_2=2`, with distinct prompts, reference images, and per-request seeds. CFG1/B2 was used as the same-shape baseline for both multi-GPU configurations.

- **CFG2/B2:** both requests completed successfully and the server logged a real `waiting=2 max_batch=2` wave. A separate traced smoke confirmed that CFG0 and CFG1 each received `num_reqs=2` with guidance rows `[[5.0, 2.0, true], [5.0, 2.0, true]]`.
- **CFG3/B2:** both requests completed successfully, the server logged a real `waiting=2 max_batch=2` wave, and CFG0, CFG1, and CFG2 each independently received the same two-request double-guidance batch.

Against CFG1/B2, both CFG2/B2 and CFG3/B2 produced:

| Request | PSNR | MAE (/255) |
|---|---:|---:|
| A (seed 12345) | 40.31 dB | 1.594 |
| B (seed 67890) | 49.82 dB | 0.312 |

Crossed-request controls were only about 7.6 dB, confirming that output rows remained mapped to the correct request. The corresponding CFG2 and CFG3 outputs were byte-identical for both requests. No runtime, NCCL, or OOM errors were observed.

This is correctness-only validation. The vGPU boxes had no direct GPU P2P, so I am not using their latency as a performance result.

**Comment by ShengleiFu** [comment]
@hsliuustc0106 @yenuo26 thank you for the review, could you help take another look again, thanks!

File: vllm_omni/diffusion/models/boogu_image/pipeline_boogu_image.py [hunk 2]
```diff
@@ -106,18 +106,15 @@ def post_process_func(images: torch.Tensor):
...
+def _boogu_batch_compatibility_key(has_reference: bool) -> tuple:
+    """Separate T2I and TI2I denoise paths.
+
+    Shared shape/guidance fields, including ``guidance_scale_2_provided``,
+    are checked by ``RequestBatchSamplingParamsKey``.
     """
     if not has_reference:
         return ("boogu_image", "t2i")
-    return ("boogu_image", "ti2i", request_id)
+    return ("boogu_image", "ti2i")
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/boogu_image/pipeline_boogu_image.py:117]: This shared TI2I key is what admits two edits into one `forward()`. Encoded instruction length is not part of the key, while `_get_instruction_feature_embeds` calls `apply_chat_template(..., padding="longest")`, so request A's VLM/DiT sequence length depends on the partner. The PR already reports MAD 2.6–6.0 for that case, which is partner coupling rather than kernel noise. Fold a length bucket (or tokenized length) into this key, or keep the shared key and add a test that actually varies padded seq_len—otherwise TI2I B>1 will change an edit whenever a longer co-arrival is scheduled beside it.

File: tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py [hunk 3]
```diff
@@ -1050,7 +1063,13 @@ def apply_chat_template(self, prompts, **kwargs):
...
         batch = len(prompts)
-        input_ids = torch.arange(batch * _SEQ_LEN, dtype=torch.long).view(batch, _SEQ_LEN)
+        input_ids = torch.zeros(batch, _SEQ_LEN, dtype=torch.long)
+        for i, messages in enumerate(prompts):
+            system_text = messages[0]["content"][0]["text"]
+            user_text = next(c["text"] for c in messages[1]["content"] if c.get("type") == "text")
+            input_ids[i, 0] = _content_id(system_text)
+            input_ids[i, 1] = _content_id(user_text)
+            input_ids[i, 2:] = torch.arange(2, _SEQ_LEN) + i
         attention_mask = torch.ones(batch, _SEQ_LEN, dtype=torch.long)
```
Comment (SamitHuang) [inline tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py:1074]: `_ImageAwareRecordingProcessor` still emits a fixed `_SEQ_LEN` and an all-ones mask, so it never exercises production `padding="longest"` / real `attention_mask`. The TI2I isolation test below then only mutates same-length partner strings, which cannot fail if A's hidden-state width or padding tokens change with B. Replace this fake with longest-padding plus a mask-aware transformer (or call the real pad path) and assert either invariance or an explicit bound; without that, `test_forward_batched_ti2i_partner_reference_isolation` does not cover the newly ungated heterogeneous-length batch.

File: tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py [hunk 4]
```diff
@@ -1541,26 +1581,212 @@ def run(prompt_a, seed_a, neg_a, prompt_b, seed_b, neg_b):
...
+        )
+        return pipeline.forward(req)[0].output
+
+    A = ("make it winter", "ugly", 1, 0.25)
+    baseline = run(*A, "a dog in a park", "blurry", 2, 0.75)
+    assert torch.equal(baseline, run(*A, "a totally different scene", "blurry", 2, 0.75))
+    assert torch.equal(baseline, run(*A, "a dog in a park", "watermark", 2, 0.75))
+    assert torch.equal(baseline, run(*A, "a dog in a park", "blurry", 999, 0.75))
+    assert torch.equal(baseline, run(*A, "a dog in a park", "blurry", 2, 0.5))
+    # Same-length partner rewrites exercise content rather than length.
```
Comment (SamitHuang) [inline tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py:1706]: This comment records that length is out of scope, but that is the mode GPU already showed partner drift. After dropping the batch=1 guard, a short vs long partner pair is a valid scheduler admission. Add a case here (once the fake encoder can pad) where B's instruction is substantially longer than A's and A's output is either unchanged or within a stated tolerance; a same-length rewrite does not substitute for that.

**Review by SamitHuang — CHANGES_REQUESTED** [review]
## Review

Ungating TI2I onto a shared `("boogu_image", "ti2i")` key matches the scheduler contract now that `guidance_scale_2_provided` is part of `RequestBatchSamplingParamsKey`. `test_ti2i_batch_key_distinguishes_explicit_image_guidance` covers the original batch=1 failure mode (equal numeric `guidance_scale_2`, different explicitness). Per-request generators / `_build_ref_latents` and the RED-arm isolation tests address cross-request mixing of prompt, seed, and reference.

The remaining merge gap is partner-dependent numerics on the path this PR newly admits. `_get_instruction_feature_embeds` uses `padding="longest"`, but the compatibility key does not include encoded length, and the CPU isolation tests pin every fake encoder to `_SEQ_LEN` with an all-ones mask. That cannot regress the coupling this PR already reports (MAD 2.6–6.0 when partner instruction lengths differ). Independent serial-vs-batched GPU checks on this head were also not pixel-exact at fixed shape (MAE ~1.7/255). Either bucket encoded length into the TI2I key, or add a CPU test whose processor actually pads like production and record the serial/B>1 quality bound in `recipes/Boogu/Boogu-Image.md` so `--max-num-seqs>1` is not treated as bit-exact vs serial.

CFG2/CFG3 × TI2I B=2 was exercised on GPU in-thread; there is still no automated check for that combination. Turbo remaining `supports_request_batch = False` is unchanged.

**Comment by ShengleiFu** [comment]
Hi @SamitHuang , thanks for the thorough review, addressed all your comments, PTAL, thanks!

**Comment by ShengleiFu** [comment]
Hi @SamitHuang , kindly ping for another review, thanks!

File: tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py [hunk 5]
```diff
@@ -1541,26 +1582,616 @@ def run(prompt_a, seed_a, neg_a, prompt_b, seed_b, neg_b):
...
+
+    class _RefContentVAE(_EditForwardVAE):
+        def encode(self, img):
+            val = float(img.float().mean().item())
+            dist = SimpleNamespace(sample=lambda generator=None: torch.full((1, 4, 8, 8), val))
+            return SimpleNamespace(latent_dist=dist)
+
+    prompt_image = PIL.Image.new("RGB", (64, 64))
+
+    def edit_prompt(text, neg, ref_fill):
```
Comment (lishunyang12) [inline tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py]: Extract these batching cases into a separate module with shared prompt and VAE helpers.

File: tests/diffusion/models/boogu_image/test_boogu_image_transformer.py [hunk 6]
```diff
@@ -604,6 +604,115 @@ def test_transformer_forward_ti2i_shape():
...
+    content, reference latent, and attention mask are held byte-identical
+    across runs; only B's padded region grows.
+    """
+    from vllm_omni.diffusion.models.boogu_image.boogu_image_transformer import (
+        BooguImageDoubleStreamRotaryPosEmbed,
+        BooguImageTransformer2DModel,
+    )
+
+    model = BooguImageTransformer2DModel(od_config=_tiny_od_config())
+    _randomize_parameters(model)
```
Comment (lishunyang12) [inline tests/diffusion/models/boogu_image/test_boogu_image_transformer.py:629]: Seed Torch before initializing and randomizing the model weights.

Comment (ShengleiFu) [reply tests/diffusion/models/boogu_image/test_boogu_image_transformer.py:629]: updated, thanks

File: tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py (hunk 5 above)
Comment (ShengleiFu) [reply tests/diffusion/models/boogu_image/test_pipeline_boogu_image.py]: updated, thanks!

**Comment by ShengleiFu** [comment]
Hi @lishunyang12 , could you ptal, just addressed both your comments, thanks!

**Comment by ShengleiFu** [comment]
Kindly ping for review @hsliuustc0106 @SamitHuang @lishunyang12 , thanks!

**Review by SamitHuang — APPROVED** [review]
LGTM
